### PR TITLE
Do not use PyObject_RichCompareBool to detect invalid comparison

### LINKED ISSUE #985: nan == nan is True when other is the same object as self
## 🐛 Bug Reports

When reporting a bug, please provide the following information. If this is not a bug report you can just discard this template.

### 🌍 Environment

 - Your operating system and version: MacOS Mojave 10.14.6
 - Your python version: 3.7.4
 - How did you install python (e.g. apt or pyenv)? Did you use a virtualenv?: virtualenv
 - Your Rust version (`rustc --version`): 1.46.0-nightly (feb3536eb 2020-06-09)
 - Your PyO3 version: 0.10
 - Have you tried using latest PyO3 master (replace `version = "0.x.y"` with `git = "https://github.com/PyO3/pyo3")?`: Yes

### 💥 Reproducing

```rust
use pyo3::prelude::*;

fn main() {
    let gil = Python::acquire_gil();
    let py = gil.python();
    let a = py.eval("float('nan')", None, None).unwrap();
    // nan == nan must not be Equal, but this test fails.
    assert_ne!(a.compare(a).unwrap(), std::cmp::Ordering::Equal);
}
```

This will fail with the below command

```
$ cargo run
```
```
thread 'main' panicked at 'assertion failed: `(left != right)`
  left: `Equal`,
 right: `Equal`', src/main.rs:8:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Here is the repo to reproduce: https://github.com/tamuhey/pyo3_nan_bug

### PR DESCRIPTION
**Pull request by kngwyu**
Fixes #985

### CODE REVIEW

File: src/types/any.rs [hunk 1]
```diff
@@ -146,36 +146,36 @@ impl PyAny {
...
-        ) -> PyResult<Ordering> {
-            let result = ffi::PyObject_RichCompareBool(a, b, ffi::Py_EQ);
-            if result == 1 {
-                return Ok(Ordering::Equal);
-            } else if result < 0 {
+            op: c_int,
+        ) -> PyResult<bool> {
+            let result = ffi::PyObject_RichCompare(a, b, op);
+            if result.is_null() {
                 return Err(PyErr::fetch(py));
```
Comment (tamuhey) [inline src/types/any.rs:158]: is it okay if you don't call `ffi::PyDECREF(result)` here?

Comment (kngwyu) [reply src/types/any.rs:158]: Oh, good catch :+1: , thank you.
I actually forgot to call it.

File: src/types/any.rs [hunk 2]
```diff
@@ -146,36 +146,25 @@ impl PyAny {
...
+        };
+        other.with_borrowed_ptr(py, |other| {
+            if do_compare(other, ffi::Py_EQ)? {
                 return Ok(Ordering::Equal);
-            } else if result < 0 {
-                return Err(PyErr::fetch(py));
             }
-            let result = ffi::PyObject_RichCompareBool(a, b, ffi::Py_LT);
-            if result == 1 {
+            if do_compare(other, ffi::Py_LT)? {
```
Comment (davidhewitt) [inline src/types/any.rs:159]: Maybe `else if do_compare(...` ?

File: src/types/any.rs [hunk 3]
```diff
@@ -146,36 +146,25 @@ impl PyAny {
...
                 return Ok(Ordering::Less);
-            } else if result < 0 {
-                return Err(PyErr::fetch(py));
             }
-            let result = ffi::PyObject_RichCompareBool(a, b, ffi::Py_GT);
-            if result == 1 {
+            if do_compare(other, ffi::Py_GT)? {
```
Comment (davidhewitt) [inline src/types/any.rs:162]: Ditto `else if` ?

File: src/types/any.rs (hunk 2 above)
Comment (kngwyu) [reply src/types/any.rs]: :+1:

File: src/types/any.rs [hunk 4]
```diff
@@ -146,36 +146,24 @@ impl PyAny {
...
-            let result = ffi::PyObject_RichCompareBool(a, b, ffi::Py_GT);
-            if result == 1 {
+            } else if do_compare(other, ffi::Py_GT)? {
                 return Ok(Ordering::Greater);
-            } else if result < 0 {
-                return Err(PyErr::fetch(py));
             }
             Err(TypeError::py_err(
                 "PyAny::compare(): All comparisons returned false",
             ))
```
Comment (davidhewitt) [inline src/types/any.rs:166]: I just realised we could change this to:

```suggestion
            } else {
                Err(TypeError::py_err(
                    "PyAny::compare(): All comparisons returned false",
                ))
            }
```

Then we could also remove the `return` keyword from all the other branches.

Comment (kngwyu) [reply src/types/any.rs]: :+1:

**Comment by pganssle** [comment]
I think this is the right thing to do in the general case, but I suspect that there's a reasonable speed-up associated with using `PyObject_RichCompareBool`, since _most_ objects actually _are_ equal when the object identity is the same.

I have not looked at this issue in any detail, and I haven't tried to do any benchmarks to see if this change even really had a major speed impact, but I'm wondering how much work it would be to either special case certain types where it is known that pointer comparison is valid, like `int`, `str`, `list`, `tuple`, `datetime.datetime`, etc, or even something more elaborate using traits (or an extra field in `PyAny` indicating whether pointer comparison is "safe" for a given ).

I would be curious to see how this change affected the speed of comparison of, say, two identical tuples or lists, since for container types this optimization is already made in `__eq__`:

```python
>>> a = [float('nan')]
>>> b = [float('nan')]
>>> a == b
False
>>> a == a
True
>>> b == b
True
```

Another possible performance regression would be two lists that are not the same object but consistent of a large number of elements all of which *are* the same object, like so:

```python
>>> OBJECT = "bbbbb"
>>> a = [OBJECT] * 10000
>>> b = [OBJECT] * 10000
>>> a == b
```

**Pull request merged**